File: 0158-Read N Characters Given Read4 II - Call multiple times/0158-Read N Characters Given Read4 II - Call multiple times.cpp
// Forward declaration of the read4 API.
int read4(char *buf);

class Solution {
public:
    /**
     * @param buf Destination buffer
     * @param n   Maximum number of characters to read
     * @return    The number of characters read
     */
    int read(char *buf, int n) {
        int i = 0;
        while (i < n) {
            if (i4 == n4) {
                i4 = 0;
                n4 = read4(buf4);
                if (n4 == 0) {
                    break;
                }
            }
            buf[i++] = buf4[i4++];
        }
        
        return i;
    }
    
private:
    char buf4[4];
    int i4 = 0, n4 = 0;
};

class Solution {
public:
    /**
     * @param buf Destination buffer
     * @param n   Maximum number of characters to read
     * @return    The number of characters read
     */
    int read(char *buf, int n) {
        int i = 0;
        while (i < n) {
            if (i4 == n4) {
                i4 = 0;
                n4 = read4(buf4);
                if (n4 == 0) {
                    break;
                }
            }
            int len = min(n4 - i4, n - i);
            memcpy(buf + i, buf4 + i4, len);
            i += len;
            i4 += len;
        }
        
        return i;
    }
    
private:
    char buf4[4];
    int i4 = 0, n4 = 0;
};

File: 0101-0200/0158-Read N Characters Given Read4 II - Call multiple times/0158-Read N Characters Given Read4 II - Call multiple times.py
"""
The read4 API is already defined for you.

    @param buf, a list of characters
    @return an integer
    def read4(buf):

# Below is an example of how the read4 API can be called.
file = File("abcdefghijk") # File is "abcdefghijk", initially file pointer (fp) points to 'a'
buf = [' '] * 4 # Create buffer with enough space to store characters
read4(buf) # read4 returns 4. Now buf = ['a','b','c','d'], fp points to 'e'
read4(buf) # read4 returns 4. Now buf = ['e','f','g','h'], fp points to 'i'
read4(buf) # read4 returns 3. Now buf = ['i','j','k',...], fp points to end of file
"""
class Solution:
    def __init__(self):
        self.buf4 = [''] * 4
        self.i4 = self.n4 = 0
    
    def read(self, buf, n):
        """
        :type buf: Destination buffer (List[str])
        :type n: Number of characters to read (int)
        :rtype: The number of actual characters read (int)
        """
        i = 0
        while i < n:
            if self.i4 == self.n4:
                self.i4 = 0
                self.n4 = read4(self.buf4)
                if self.n4 == 0:
                    break
            num = min(self.n4 - self.i4, n - i)
            for k in range(num):
                buf[i + k] = self.buf4[self.i4 + k]
            i += num
            self.i4 += num
        return i
